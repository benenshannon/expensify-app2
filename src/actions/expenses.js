import uuid from 'uuid';
import database from '../firebase/firebase';

//How the action generators work:
//component calls action generator
//action generator returns an object
//component dispatches the object
//redux store changes

//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
        const expense = {description, note, amount, createdAt};

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSE
export const removeExpense = (
    { id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//startRemoveExpense
export const startRemoveExpense= ({ id } = {}) => {
    return (dispatch) => {
        //dispatch gets passed to this function by the redux library
        return database.ref(`expenses/${id}`).remove().then(() => {
            //once the item has been removed, the removeExpense action
            //generator is then called
            dispatch(removeExpense( { id }));
        });
    };
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// manipulate the redux store
//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//THIS WILL ACTUALLY FETCH THE DATA
//export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
};