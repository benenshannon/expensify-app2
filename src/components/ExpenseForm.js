import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
//use the faq on https://github.com/airbnb/react-dates to help understand
//what is needed to use SingleDatePicker

// using date elements example: console.log(new Date().getFullYear()+1);
//moment is the standard approach, easier than date
const now = moment();
console.log(now.format('MMM Do, YYYY',));

export default class ExpenseFrom extends React.Component {
    constructor(props) {
        super(props);
        this.state = {   
         //"state" sets the current state    
         description: props.expense ? props.expense.description : '',
         note: props.expense ? props.expense.note : '',
         amount: props.expense ? (props.expense.amount / 100).toString() : '',
         createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
         calendarFocused: false,
         error: ''
        };
    }

//onDescriptionChange is the arrow function that creates a 
//const (description) getting the value from e.target.value (e = event)
//this.setState is calling the description value set in the 
//const description code immediately above
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
        this.setState(() => ({amount}));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
        this.setState(() => ( { createdAt }));
        }
        //this prevents the user from deleting the date from the input field
    };

    onFocusChange = ( { focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        //this prevents a full page refresh
        e.preventDefault();
        //'if' will check if there is no info in description or amount
        if (!this.state.description || !this.state.amount) {
            //set the updater function
            this.setState(() => ({ 
                error: 'Please provide description and amount'})); 
        } else {
            this.setState(() => ({error: ''}));
        this.props.onSubmit({
            //this is creating the properties of the onSubmit object
            //to be imported into the AddExpensePage
            description: this.state.description,
            amount: parseFloat(this.state.amount, 10) * 100,
            createdAt: this.state.createdAt.valueOf(),
            note: this.state.note
        });
        }
    };

    render() {
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit= {this.onSubmit} >
              <input 
                type = "text"
                placeholder = "Description"
                autoFocus 
                value = {this.state.description}
                //this.state calls the above state (current state)
                onChange={this.onDescriptionChange}
                />    
              <input
                type = "text"
                placeholder = "Amount"
                value = {this.state.amount} 
                onChange = {this.onAmountChange}/>
              <SingleDatePicker
                date={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.calendarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
            //numberOfMonths etc is in the online documentation in react-dates
              />
              <textarea
                placeholder = "Add a note for your expense (optional)"
                value = {this.state.note}
                onChange={this.onNoteChange}>
                </textarea>
              <button>
                Add expense
              </button>
            </form>
            </div>
        )
    };
};

//Use local component state to track the changes & only dispatch data to store
//when the button is clicked

