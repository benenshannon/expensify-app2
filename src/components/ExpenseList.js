import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />;
            })
            )
        }
    </div>
);

//mapping the state (redux store) to the props
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//HOC
export default connect(mapStateToProps)(ExpenseList);

//This file is referenced in the components/ExpenseDashboard.js file