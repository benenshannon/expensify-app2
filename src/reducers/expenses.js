// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState , action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //.filter is a way of searching an array for
            //objects that meet a criteria - such as !== action.id
            // and returning the new filtered array 
            return state.filter(( { id }) => id !== action.id);
            //return state.filter(( { id }) => {return id !== action.id});
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                };
            });
        case 'SET_EXPENSES':
        // returning the action object with the expense array attached
            return action.expenses;   
        default:
            return state;
    }
};