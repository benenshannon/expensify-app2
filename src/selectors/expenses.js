import moment from 'moment';

//Get visible expenses
export default (expenses, {text, sortBy, startDate, endDate }) => {
    //arrow function calling "expenses" array, and 
    //"filters" (destructured by using {})
        return expenses.filter((expense) => {
            const createdAtMoment = moment(expense.createdAt)
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true ;
           const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true ;
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
            return startDateMatch && endDateMatch && textMatch;
        }).sort((a, b) => {
            if(sortBy === 'date') {
                return a.createdAt < b.createdAt ? 1 : -1;
                //if a is less than than b, then true = 1, meaning b will come first
            }
            else if(sortBy === 'amount') {
                return a.amount < b.amount ? 1 : -1;
            }
        });
    }; 
    