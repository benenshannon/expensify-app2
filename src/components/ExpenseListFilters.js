import React from 'react';
import {connect} from 'react-redux';
//connect connects to filters and expenses from the configureStore store
//it does this by the command "props.filters/expenses"
import {DateRangePicker} from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    //as this is now a class, whenever "props.filters/expenses" is called 
    //it needs to be preceeded by "this".props.filters/expenses
    //(as "this" refers to the props within the class)
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => {
        //dispatch to the store the setStartDate action with startDate as 
        //selected
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() =>({ calendarFocused }));
    };

    onTextChange = (e) =>{
        //"e" is the event change
        //dispatch sends info to the store
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if(e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
             this.props.sortByAmount();
        }
    };

    render() {
        return (
            <div>
            <input 
            type="text" 
            value = {this.props.filters.text} 
                onChange={this.onTextChange}
                /> 
            <select
            //inside the <select> command is where we identify the 'value' item
            //this value is then called in the <option> code
            //props.filters (and props.expenses) - this is calling the export from
            //.reducers/filters.js, as props calls the original store (configureStore)
               value={this.props.filters.sortBy}
               onChange={this.onSortChange}      
               >
              <option value ="amount">Amount</option>
              <option value ="date">Date</option>
            </select>
            <DateRangePicker
            //startDate gets it's value off the filters object (this is the 
            //filters file in the reducers folder, imported into 
            //configureStore and then called using import {connect})
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              //on dates change doesnt exist anywhere (hence no this.props
              //so it gets created in the class, above)
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths = {1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
            </div>
        );
    };
};

//mapStateToProps calls the state we want, and then is entered in the export
//default command
const mapStateToProps = (state) => ({filters: state.filters});


const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

//export connected version of ExpenseListFilters
//first brackets define what we want, second brackets define what we export
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);