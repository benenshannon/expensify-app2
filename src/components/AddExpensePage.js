import React from 'react';
import { connect } from 'react-redux';
//adding connect so that the expenseform can be submitted to the store
import ExpenseForm from './ExpenseForm';
//ExpenseForm is a class
import { startAddExpense } from '../actions/expenses';
//this imports addExpense from the actions folder with all the properties
//(description, amount, createdAt etc)

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
    <div>
    <h1>Add Expense</h1>
    <ExpenseForm 
      onSubmit={this.onSubmit}
      />
  </div>
);
  }
};


//CODE BELOW IS SUPERCEDED BY BY THE CODE STARTING ON ROW 10 ABOVE
//const AddExpensePage = (props) => (
//<div>
//  <h1>Add Expense</h1>
//  <ExpenseForm 
//    onSubmit={(expense) => {
   //onSubmit prop, gets called when form is properly submitted and calls the 
   //'expense' argument
//   props.onSubmit(expense);
//   props.history.push('/');
  //props.history.push uses queries that exist in URL/JSX language already
//    }}
//  />
//</div>
//As ExpenseForm is JSX already, it needs be called as <ExpenseForm />
//);

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);