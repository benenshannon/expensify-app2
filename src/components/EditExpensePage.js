import React from 'react';
import  { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <ExpenseForm 
         expense={this.props.expense}
         onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
};


//const EditExpensePage = (props) => {
//  return  (
//    <div>
//      <ExpenseForm 
      //this returns the ExpenseForm js
//      expense={props.expense}
//      onSubmit = {(expense) => {
        //dispatch the action to edit expense
        // redirect to the dashboard page
//        props.dispatch(editExpense(props.expense.id, expense));
        //editExpense has two arguments - id and updates, so get
        //id from the existing props.expense, and then the other updates 
        //from existing 'expense' folder
//        props.history.push('/');
//      }}
//      />
//      <button onClick={(e) => {
//        props.dispatch(removeExpense({id: props.expense.id}));
//        props.history.push('/');
//      }} >Remove</button>
//    </div>
//  );
//};

const mapStateToProps = (state, props) => ({
  //take some of the props from the HOC (higher order component, which is State)
    expense: state.expenses.find((expense) =>
      expense.id === props.match.params.id)
    });
    //.find lets us find exact matches within an array

const mapDispatchToProps = (dispatch, props) => ({
   editExpense: (id, expense) => dispatch(editExpense(id, expense)),
   startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);