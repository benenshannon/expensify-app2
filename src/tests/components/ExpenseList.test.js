import React from 'react';
import { shallow } from 'enzyme';
import "../setupTests";
import { ExpenseList } from '../../components/ExpenseList';
//ExpenseList is in {} as it is a named export (as there is also
//a default export in the components/ExpenseList.js file)
import expenses from '../fixtures/expenses';

//Test the unconnected ExpenseList - so we dont want to get this from the   
//store (configstore)

test('should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    //ExpenseList expects the expenses property, so that is what is added
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list with empty message', () =>{
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    //ExpenseList expects the expenses property, so that is what is added
    expect(wrapper).toMatchSnapshot();
});