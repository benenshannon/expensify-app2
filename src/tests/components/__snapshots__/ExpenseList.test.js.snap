// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={189500}
    createdAt={0}
    description="rent"
    id="1"
    key="1"
    note="this was the old rent"
  />
  <ExpenseListItem
    amount={230000}
    createdAt={-345600000}
    description="mortgage"
    id="2"
    key="2"
    note="this was the first mortgage"
  />
  <ExpenseListItem
    amount={250000}
    createdAt={345600000}
    description="new mortgage"
    id="3"
    key="3"
    note="this was the new mortgage payment"
  />
</div>
`;
